### ISSUE TITLE
DefaultLwM2mNodeEncoder uses too much memory

### ISSUE THREAD

**Issue by mjjaniec** [eclipse/leshan#331]
Hi!

Use case:
I was testing leshan with Anjay (https://github.com/AVSystem/Anjay)
I was tring to send 100MB OPAQUE value via leshan, but it took all the memory and exploded.

Problem:
`DefaultLwM2mNodeEncoder` uses to much memory

Solution:
In `DefaultLwM2mNodeEncoder` at line 79 there is:

        LOG.trace("Encoded node {}: {}", node, Arrays.toString(encoded));

which should be changed to:

        if (LOG.isTraceEnabled()) {
            LOG.trace("Encoded node {}: {}", node, Arrays.toString(encoded));
        }

Explanation: 
The problem is that `Arrays.toString(byte[] a)` is quite heavy operation. 
I've prepared sample application to check it out:

```
import java.util.Arrays;

public class Main {

    public static void main(String[] args) {

        int size = 100 * 1024 * 1024;  // 100 MB
        byte[] bytes = new byte[size];
        Arrays.fill(bytes, (byte)-111);
        
        String x = Arrays.toString(bytes);
        System.out.println(x.substring(0, 10));
    }
}
```

On my PC, this application throws Out of Memory exception with `-Xmx5G`, and passes with `-Xmx6G`.
Quite a lot.

**Comment by jvermillard** [comment]
I wonder why you want to send a 100MB file over something like CoAP which is unable to saturate the bandwidth like TCP, let's imagine your use case is legit :smile:.

In place of
```
if (LOG.isTraceEnabled()) {
    LOG.trace("Encoded node {}: {}", node, Arrays.toString(encoded));
}
```

You can try:
```
LOG.trace("Encoded node {}: {}", node, (Object)encoded);
```
The logging shoud do the arrays to string only if the log level is trace.

**Comment by mjjaniec** [comment]
I was testing capabilities, actually it's not about sending 100MB but about using 20 - 40 times more memory than is needed.

**Comment by jvermillard** [comment]
Yes. I agree it must be fixed.

**Comment by sbernard31** [comment]
It should be fixed on master (commit 82c568cd03c463b9221eb3424474325138449e55)
Thx to reporting this @mjjaniec :)
